### ISSUE TITLE
[QUESTION]Memory leak occurs when createSignalingClientSync() is executed in multiple threads and fails.

### ISSUE THREAD

**Issue by pipsnh02** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1194]
（Version：Release 1.5.0 of the Amazon Kinesis Video WebRTC C SDK）

When createSignalingClientSync() runs multithreaded and fails
The memory allocated in the API has not been released.
※Confirmed by VmRSS value of process information

If createSignalingClientSync () fails repeatedly for 1 minute, memory usage will increase by about 400K.
(The function is executed 46 times in 1 minute)

A state in which you cannot intentionally connect to the signaling server (a state without LAN).
Execute handle generation API with multiple threads (no exclusion)

Main Thread
　└Sub Thread 1～16 

The main thread runs initKvsWebRtc()
The subthread creates a handle as a master and connects to each signaling channel.
If the connection failed, I put a 500ms wait and ran the API again.

Question ①：
A memory leak has occurred when createSignalingClientSync() is run multithreaded and fails.
Is there any procedure the user should take if this API fails?

Question ②：
I am running createSignalingClientSync() in multiple threads.
Is this API thread-safe?

Question ③：
The parameters of createSignalingClientSync() are listed below.
Are there any other required parameters?

・ChannelInfo										
	channelInfo.version　 = CHANNEL_INFO_CURRENT_VERSION;
	channelInfo.pKmsKeyId	= NULL;
	channelInfo.tagCount	= 0;
	channelInfo.pTags= NULL;
	channelInfo.channelType		= SIGNALING_CHANNEL_TYPE_SINGLE_MASTER;
	channelInfo.channelRoleType	= SIGNALING_CHANNEL_ROLE_TYPE_MASTER;
	channelInfo.cachingPolicy	= SIGNALING_API_CALL_CACHE_TYPE_DESCRIBE_GETENDPOINT;
	channelInfo.cachingPeriod	= SIGNALING_API_CALL_CACHE_TTL_SENTINEL_VALUE;
	channelInfo.asyncIceServerConfig = FALSE;
	channelInfo.retry		= TRUE;
	channelInfo.reconnect		= TRUE;
	channelInfo.pCertPath		= KVS_CA_CERT_PATH;
	channelInfo.messageTtl		= 0;
	channelInfo.pRegion		= DEF_REGION;

・SignalingClientInfo											
	clientInfo.version		= SIGNALING_CLIENT_INFO_CURRENT_VERSION;
	clientInfo.loggingLevel		= LOG_LEVEL_SILENT;
	clientInfo.cacheFilePath	= NULL;
	clientInfo.clientId		= "client"

**Comment by MushMal** [comment]
Looking at your parameters I don't see anything that is immediately standing out - they look OK.

Create API is deeply isolated so it should be OK to run multiple instances in parallel as long as they don't share your provided objects. The only caveat is the platform and software dependencies. As the signaling client relies on libWebSockets, in turn it relies on OpenSSL and the OS sockets. Any implementation that is not properly isolated would cause an issue.

I would strongly recommend using the stock sample code on a commodity platform (say Linux on PC). Then, modify the master sample code to add a loop when it creates the signaling channel and ensure the cleanup clause is included in the loop. Run it without network connection and check whether you see any leaks. There are multiple unit tests that test this scenario under the ASAN which would flag a leak but perhaps with a real network outage there is a separate path in the dependencies that might cause a leak - not sure.

**Comment by tencho0221** [comment]
>There are multiple unit tests that test this scenario under the ASAN which would flag a leak but perhaps with a real network outage there is a separate path in the dependencies that might cause a leak - not sure.

Sorry to interrupt the conversation.
There seems to be a similar problem with libwebsockets. Is it related?
https://github.com/warmcat/libwebsockets/issues/2044

**Comment by pipsnh02** [comment]
MushMal
thank you.
I modified the sample source as suggested and checked it, but a memory leak occurred.
It was carried out in a single thread.

Does the other dependent path refer to [libwebsockets # 2044](https://github.com/warmcat/libwebsockets/issues/2044)?
Please let me know if there are any other paths related to leaks.

Is there a sample that provides multiple masters in one process?

**Comment by pipsnh02** [comment]
tencho0221
thank you.
This time I used libWebSocket v3.2.3.
Since the leaked part described in libwebsockets # 2044 existed
I uploaded it to the latest branch (libwebsockets = v4.2.1) and tried it.
It was carried out in a single thread.
The amount of leak has decreased, but it is still leaking.

I want to isolate whether the remaining leak is a KVS problem or a libwebsockets problem.
Is there an effective means?

**Comment by MushMal** [comment]
Nothing comes to my mind off the top of my head. The object graph is isolated which means that the signaling client object itself is self-contained and can have multiple copies within the execution unit. Some of the unit tests do have multiple signaling clients I believe. You should check out the rest of the system - for example, the callbacks going into the signaling client and other auxiliary structures like auth, etc.. Try to start from something minimal - have two signaling clients and tear them down and make sure no leaks are present. You can check out the signaling client unit tests.

**Comment by disa6302** [comment]
@pipsnh02 , 

Going through the code base, I do not see a path for potential thread leaks invoking `createSignalingClientSync`. Is there a way for you to get info on the threads on the system? That way we can look into pinpointing the issue.

**Comment by disa6302** [comment]
Closing due to no update. Feel free to reopen if this is still an issue.

**Comment by pipsnh02** [comment]
After applying two patches to libWebSockets v3.2.3, the leak disappeared.
① [#2178 patch](https://github.com/warmcat/libwebsockets/commit/ffa0a38a1a721fc1ec96ad21916438846069bd77.patch)
② [#1190 ](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/issues/1190)
（Only some leaks were solved with ① alone, and all were solved with ① ＋ ②）

I think ① is a leak repair,
②is a patch for the purpose of solving another issue (the case where the connection to the signaling server ends with a timeout error).

I don't understand why the patch ② helps eliminate the leak.
Do you know why?

**Comment by ChrisZhangJin** [comment]
hmm.... i guess [#2178 patch](https://github.com/warmcat/libwebsockets/commit/ffa0a38a1a721fc1ec96ad21916438846069bd77.patch) was for [the issue i created in 2020 #2044](https://github.com/warmcat/libwebsockets/issues/2044)
